File: app/src/main/java/cn/ifafu/ifafu/mvp/base/BaseActivity.java
package cn.ifafu.ifafu.mvp.base;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import cn.ifafu.ifafu.mvp.base.i.IPresenter;
import cn.ifafu.ifafu.mvp.base.i.IView;

public abstract class BaseActivity<P extends IPresenter> extends AppCompatActivity implements IView {

    protected P mPresenter;
    protected final String TAG = this.getClass().getSimpleName();
    private Unbinder unbinder;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (initLayout(savedInstanceState) != 0) {
            setContentView(initLayout(savedInstanceState));
        }
        unbinder = ButterKnife.bind(this);
        initData(savedInstanceState);
        if (mPresenter != null) {
            mPresenter.onCreate();
        }
    }

    public int initLayout(@Nullable Bundle savedInstanceState) {
        return 0;
    }

    public abstract void initData(@Nullable Bundle savedInstanceState);

    @Override
    public void showMessage(String msg) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(int stringRes) {
        Toast.makeText(this, stringRes, Toast.LENGTH_SHORT).show();
    }

    @Override
    public Activity getActivity() {
        return this;
    }

    @Override
    public Context getContext() {
        return getApplicationContext();
    }

    @Override
    public void showLoading() {
    }

    @Override
    public void hideLoading() {
    }

    @Override
    public void openActivity(Intent intent) {
        startActivity(intent);
    }

    @Override
    public void killSelf() {
        finish();
    }

    @Override
    protected void onDestroy() {
        hideLoading();
        if (mPresenter != null) {
            mPresenter.onDestroy();
            mPresenter = null;
        }
        if (unbinder != null) {
            unbinder.unbind();
        }
        super.onDestroy();
    }

}

File: app/src/main/java/cn/ifafu/ifafu/base/mvvm/BaseActivity.kt
package cn.ifafu.ifafu.base.mvvm

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import cn.ifafu.ifafu.app.Constant
import cn.ifafu.ifafu.mvp.login.LoginActivity
import cn.ifafu.ifafu.view.dialog.LoadingDialog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

abstract class BaseActivity<VDB : ViewDataBinding, VM : BaseViewModel> : AppCompatActivity(), UIEvent {

    protected open val loadingDialog by lazy {
        LoadingDialog(this).apply {
            setText("加载中")
            setCancelable(true)
        }
    }

    protected val mBinding: VDB by lazy {
        DataBindingUtil.setContentView(this, getLayoutId()) as VDB
    }

    protected lateinit var mViewModel: VM

    protected abstract fun getLayoutId(): Int

    protected abstract fun getViewModel(): VM?

    protected abstract fun initActivity(savedInstanceState: Bundle?)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding.lifecycleOwner = this
        getViewModel()?.run {
            this.event = this@BaseActivity
            mViewModel = this
        }
        initActivity(savedInstanceState)
    }

    override fun onDestroy() {
        mBinding.unbind()
        super.onDestroy()
    }

    override suspend fun showMessage(message: String) = withContext(Dispatchers.Main) {
        Toast.makeText(this@BaseActivity, message, Toast.LENGTH_SHORT).show()
    }

    override suspend fun showDialog() = withContext(Dispatchers.Main) {
        loadingDialog.show()
    }

    override suspend fun hideDialog() = withContext(Dispatchers.Main) {
        loadingDialog.cancel()
    }

    override suspend fun startLoginActivity() = withContext(Dispatchers.Main) {
        startActivityForResult(Intent(this@BaseActivity, LoginActivity::class.java), Constant.ACTIVITY_LOGIN)
        finish()
    }

    override suspend fun finishIt() = withContext(Dispatchers.IO) {
        finish()
    }

}